pallets/creators-staking/src/types.rs

```diff
@@ -0,0 +1,417 @@
+    /// * `stake(7, 100)` will result in `[<5, 1000>, <7, 1400>]`
+    /// * `stake(9, 200)` will result in `[<5, 1000>, <7, 1400>, <9, 1600>]`
+    pub(super) fn stake(&mut self, current_era: EraIndex, value: Balance) -> Result<(), &str> {
+        if let Some(era_stake) = self.stakes.last_mut() {
+            if era_stake.era > current_era {
+                return Err("Unexpected era")
+            }
+
+            let new_stake_value = era_stake.staked.saturating_add(value);
+
+            if current_era == era_stake.era {
+                *era_stake = EraStake::new(new_stake_value, current_era)
+            } else {
+                self.stakes.push(EraStake::new(new_stake_value, current_era))
+            }
+        } else {
+            self.stakes.push(EraStake::new(value, current_era));
+        }
+
+        Ok(())
+    }
+
+    /// Unstakes some value in the specified era.
+    ///
+    /// User should ensure that given era is either equal or greater than the
+    /// latest available era in the staking info.
+    ///
+    /// # Example 1
+    ///
+    /// `stakes: [<5, 1000>, <7, 1300>]`
+    /// * `unstake(7, 100)` will result in `[<5, 1000>, <7, 1200>]`
+    /// * `unstake(9, 400)` will result in `[<5, 1000>, <7, 1200>, <9, 800>]`
+    /// * `unstake(10, 800)` will result in `[<5, 1000>, <7, 1200>, <9, 800>, <10, 0>]`
+    ///
+    /// # Example 2
+    ///
+    /// `stakes: [<5, 1000>]`
+    /// * `unstake(5, 1000)` will result in `[]`
+    ///
+    /// Note that if no unclaimed eras remain, vector will be cleared.
+    pub(super) fn unstake(&mut self, current_era: EraIndex, value: Balance) -> Result<(), &str> {
+        if let Some(era_stake) = self.stakes.last_mut() {
+            if era_stake.era > current_era {
+                return Err("Unexpected era")
+            }
+
+            let new_stake_value = era_stake.staked.saturating_sub(value);
+            if current_era == era_stake.era {
+                *era_stake = EraStake::new(new_stake_value, current_era)
+            } else {
+                self.stakes.push(EraStake::new(new_stake_value, current_era))
+            }
+
+            // Removes unstaked values if they're no longer valid for comprehension
+            if !self.stakes.is_empty() && self.stakes[0].staked.is_zero() {
+                self.stakes.remove(0);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// `Claims` the oldest era available for claiming.
+    /// In case valid era exists, returns `(claim era, staked amount)` tuple.
+    /// If no valid era exists, returns `(0, 0)` tuple.
+    ///
+    /// # Example
+    ///
+    /// The following example will demonstrate how the internal vec changes when `claim` is called
+    /// consecutively.
+    ///
+    /// `stakes: [<5, 1000>, <7, 1300>, <8, 0>, <15, 3000>]`
+    ///
+    /// 1. `claim()` will return `(5, 1000)`
+    ///     Internal vector is modified to `[<6, 1000>, <7, 1300>, <8, 0>, <15, 3000>]`
```

Comment:
Just look into a code.

This function claims for just 1 era.
In the initial vector, there was a record for 5th era. We claim for the 5th era, and we don't just delete the record, because there's still a stake for the 6th era, so we increase an era in the record to be able to know on the next claim that we still need to claim for the 6th era.

Reply:
6 is the next unclaimed era, and not to skip it when claiming, we do inherit stake from the claiming era if the difference with the next one is greater than 1.

Reply:
Would be nice to improve docs a lil bit. How about this example?
<img width="509" alt="image" src="https://github.com/dappforce/subsocial-parachain/assets/153928/56a2d9a6-7daa-4dd9-8642-ec25d279c679">